### PR TITLE
feat(lineage) Adds ability to edit lineage manually from the UI

### PR DESCRIPTION
**Pull request by chriscollins3456**
This PR adds the ability for users to edit lineage for entities manually through the UI. Right now, we only support manual lineage on the entity level and for the four core entities: Datasets, Charts, DataJobs, and Dashboards. In order to edit lineage through the UI, a user must either have metadata privileges for `EDIT_LINEAGE` on the two entities being affected (either an edge being added or an edge being removed) or superuser `EDIT_ENTITY` privileges on both entities.

You can edit lineage through the UI on both the lineage visualization as well as the Lineage tab on an entity page. The new GraphQL endpoint accepts a list of `edgesToAdd` and a list of `edgesToRemove`. Both list of edges consist of `LineageEdge` objects which look like this:
```
input LineageEdge {

    downstreamUrn: String!

    upstreamUrn: String!
}
```

This PR also introduces a new `Edge` record type so we can keep track of created and updated audit stamps as well as some new `properties` that allow us to store that information on the graph edge itself. We can then get and filter by these new fields. Specifically in the PR we're getting the created actor and time and displaying that in the UI if the actor is real to show info for users to keep track of who and when edges are being added.

Here's some screenshots:
image

image

image

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-web-react/src/Mocks.tsx [hunk 1]
```diff
@@ -142,7 +142,8 @@ export const dataset1 = {
             displayName: 'HDFS',
             type: PlatformType.FileSystem,
             datasetNameDelimiter: '.',
-            logoUrl: '',
+            logoUrl:
+                'https://raw.githubusercontent.com/datahub-project/datahub/master/datahub-web-react/src/images/lookerlogo.png',
```
Comment (jjoyce0510) [inline datahub-web-react/src/Mocks.tsx:146]: is this change intended? cannot tell

File: datahub-web-react/src/app/entity/EntityRegistry.tsx [hunk 2]
```diff
@@ -128,6 +128,10 @@ export default class EntityRegistry {
                         entity: relationship.entity as EntityInterface,
                         type: (relationship.entity as EntityInterface).type,
                     })),
+                downstreamRelationships: genericEntityProperties?.downstream?.relationships
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/EntityRegistry.tsx:131]: minor: we should probably handle this filtering at the backend layer if we do not

File: datahub-web-react/src/app/entity/EntityRegistry.tsx [hunk 3]
```diff
@@ -128,6 +128,10 @@ export default class EntityRegistry {
...
+                    ?.filter((relationship) => relationship.entity)
+                    // eslint-disable-next-line @typescript-eslint/dot-notation
+                    ?.filter((relationship) => !relationship.entity?.['status']?.removed),
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/EntityRegistry.tsx:134]: I'm assuming this is safe even when 'status' doesn't exist due to the ?

File: datahub-web-react/src/app/entity/chart/ChartEntity.tsx [hunk 4]
```diff
@@ -100,14 +100,6 @@ export class ChartEntity implements Entity<Chart> {
                     properties: {
                         defaultDirection: LineageDirection.Upstream,
                     },
-                    display: {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/entity/chart/ChartEntity.tsx:103]: why remove this?

Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/chart/ChartEntity.tsx:103]: basicaly lineage tab will always be visible now? got it. that makes sensee. forget the above!

File: datahub-web-react/src/Mocks.tsx (hunk 1 above)
Comment (chriscollins3456) [reply datahub-web-react/src/Mocks.tsx:146]: yeah - this is just for mocks to check if a logo url is there

File: datahub-web-react/src/app/entity/EntityRegistry.tsx (hunk 3 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/EntityRegistry.tsx:134]: indeed

File: datahub-web-react/src/app/lineage/__tests__/LineageEdges.test.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import { dataset1, dataset4WithLineage } from '../../../Mocks';
+import { getTestEntityRegistry } from '../../../utils/test-utils/TestPageContainer';
+import LineageEdges from '../manage/LineageEdges';
+import { Direction } from '../types';
+
+const mockEntityRegistry = getTestEntityRegistry();
+jest.mock('../../useEntityRegistry', () => ({
+    useEntityRegistry: () => mockEntityRegistry,
+}));
+
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/__tests__/LineageEdges.test.tsx:12]: nice tests!!

File: datahub-web-react/src/app/lineage/__tests__/constructTree.test.ts [hunk 6]
```diff
@@ -397,6 +407,64 @@ describe('constructTree', () => {
                     status: null,
                     platform: airflowPlatform,
                     subtype: undefined,
+                    canEditLineage: false,
+                },
+            ],
+        });
+    });
+
+    it('should construct a tree taking into account updatedLineages in state', () => {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/__tests__/constructTree.test.ts:416]: This whole updated lineages pattern is a bit tricky to grasp conceptually

File: datahub-web-react/src/app/entity/EntityRegistry.tsx (hunk 2 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/EntityRegistry.tsx:131]: yeah I'm pretty sure we do, but I'm just mimicking the existing `downstreamChildren` above to be safe

File: datahub-web-react/src/app/lineage/__tests__/manageLineage.test.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
+import { dataFlow1, dataJob1, dataset1, dataset2, dataset3 } from '../../../Mocks';
+import { existsInEntitiesToAdd } from '../manage/AddEntityEdge';
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/__tests__/manageLineage.test.tsx:2]: nice component test!

File: datahub-web-react/src/app/lineage/manage/LineageEdges.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                    <Empty description={`No ${lineageDirection.toLocaleLowerCase()} entities`} />
+                </EmptyWrapper>
+            )}
+            {filteredRelationships &&
+                filteredRelationships?.map((relationship) =>
+                    relationship.entity ? (
+                        <EntityEdge
+                            key={relationship.entity.urn}
+                            entity={relationship.entity}
+                            removeEntity={removeEntity}
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/manage/LineageEdges.tsx:80]: any chance actor is undefined? could this cast cause problems?

File: datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            <UnderlineWrapper onClick={centerEntity}>Focus</UnderlineWrapper> on this entity to make {direction} edits.
+        </div>
+    );
+}
+
+function getDownstreamDisabledPopoverContent(canEditLineage: boolean, isDashboard: boolean, centerEntity?: () => void) {
+    if (!canEditLineage) {
+        return UNAUTHORIZED_TEXT;
+    }
+    if (isDashboard) {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx:36]: nice!

File: datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+                                                ) : (
+                                                    <PopoverContent centerEntity={centerEntity} direction="upstream" />
+                                                )
+                                            }
+                                            overlayStyle={
+                                                isUpstreamDisabled ? { zIndex: POPOVER_Z_INDEX } : { display: 'none' }
+                                            }
+                                        >
+                                            <div>
+                                                <Menu.Item
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx:112]: minor: It would have been nice to factor some of these items out. maybe in a follow up

File: datahub-web-react/src/app/lineage/__tests__/constructTree.test.ts (hunk 6 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/lineage/__tests__/constructTree.test.ts:416]: yeah agreed but tbh my hands were kind of tied with how this lineage viz renders things..

File: datahub-web-react/src/app/lineage/manage/UserAvatar.tsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+interface Props {
+    createdActor?: CorpUser | null;
+    createdOn?: number | null;
+}
+
+export default function UserAvatar({ createdActor, createdOn }: Props) {
+    const entityRegistry = useEntityRegistry();
+    const avatarPhotoUrl = createdActor?.editableProperties?.pictureLink;
+    const userName = entityRegistry.getDisplayName(EntityType.CorpUser, createdActor);
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/manage/UserAvatar.tsx:40]: just want to confirm, if createdActor is not defined, this should still work?

File: datahub-web-react/src/app/lineage/manage/LineageEdges.tsx (hunk 8 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/lineage/manage/LineageEdges.tsx:80]: yeah it could be or null, but the `EntityEdge` is expecting that it could be undefined or null as well. I just needed to type case because if it does exist, it's a `CorpUser` so we can access the appropriate fields in `EntityEdge`

File: datahub-web-react/src/app/lineage/manage/UserAvatar.tsx (hunk 11 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/lineage/manage/UserAvatar.tsx:40]: userName would basically be undefined in that case

File: datahub-web-react/src/app/lineage/utils/constructTree.ts [hunk 12]
```diff
@@ -1,16 +1,87 @@
...
     fetchedEntities: FetchedEntities,
     direction: Direction,
     entityRegistry: EntityRegistry,
+    updatedLineages: UpdatedLineages,
 ): NodeData {
     if (!entityAndType?.entity) return { name: 'loading...', children: [] };
     const constructedNodes = {};
 
+    let updatedFetchedEntities = fetchedEntities;
+    Object.entries(updatedFetchedEntities).forEach((entry) => {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/utils/constructTree.ts:65]: this block is difficult to follow.. as is as tjhe code in the file, unfortunately.

File: datahub-web-react/src/app/lineage/manage/UserAvatar.tsx (hunk 11 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/lineage/manage/UserAvatar.tsx:40]: yup - i do a check above to not even render this component if there is no actor

File: datahub-web-react/src/app/lineage/utils/manageLineageUtils.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    if (lineageDirection === Direction.Downstream) {
+        edgesToAdd = entitiesToAdd.map((entity) => ({ upstreamUrn: entityUrn, downstreamUrn: entity.urn }));
+        edgesToRemove = entitiesToRemove.map((entity) => ({ upstreamUrn: entityUrn, downstreamUrn: entity.urn }));
+    }
+
+    return { edgesToAdd, edgesToRemove };
+}
+
+export function getValidEntityTypes(lineageDirection: Direction, entityType?: EntityType) {
```
Comment (jjoyce0510) [inline datahub-web-react/src/app/lineage/utils/manageLineageUtils.ts:68]: i really wish there were a more generic way to do this.. not sure there is right now..

File: datahub-web-react/src/app/lineage/manage/LineageEdges.tsx (hunk 8 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/lineage/manage/LineageEdges.tsx:80]: awesome

File: datahub-web-react/src/app/lineage/utils/constructTree.ts (hunk 12 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/lineage/utils/constructTree.ts:65]: yeah :( i was really struggling in here - as mentioned our lineage code can use a face lift

File: datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx (hunk 10 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/lineage/manage/ManageLineageMenu.tsx:112]: agreed - it got a good amount busier at the last leg of this feature with adding more and more to the menu - a little cleanup wouldn't hurt anybody
